### PR TITLE
Add custom header support for HTTP proxy

### PR DESCRIPTION
**Pull request by PMExtra**
# Background

I'm using `Xray` as a general network tool without pass GFW purpose.

In my case, I made all network traffic from the private subnet be forwarded to Xray with TPROXY.

Then Xray will do some sniffing to override the destination, then forward the request to a Squid service by HTTP Proxy protocol.

Then I need to take access controlling and auditing with Squid.

So I have to know what's the original source IP address.

Fortunately, Squid support solve source address from `X-Forwarded-For` header which send with `CONNECT` proxy request.

# Changes

So I made this change to support append custom headers with HTTP PROXY:

I added a new optional setting `headers` in `OutboundConfigurationObject` of HTTP outbound.

It support [`text/template`](https://pkg.go.dev/text/template) format.

For now, there are 2 properties can be accessed: `Source net.Destination` and `target net.Destination`.

There is an configuration example:

```json
{
  "outbounds": [
    {
      "protocol": "http",
      "settings": {
        "servers": [
          {
            "address": "proxy.my.domain",
            "port": 8080
          },
          "headers": {
            "X-Forwarded-For": "{{.Source.Address}}",
            "X-Forwarded-Host": "my.domain",
            "X-Real-IP": "{{if .Source.Address.Family.IsIP}}{{.Source.Address}}{{end}}",
            "Forwarded": "by=4.3.2.1;for={{if .Source.Address.Family.IsIPv4}}{{.Source.Address}}{{else if .Source.Address.Family.IsIPv6}}\"{{.Source.Address}}\"{{end}};host=my.domain;proto=http"
          }
        ]
      }
    }
  ]
}
```

> According to https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Forwarded , ipv6 should be quoted and enclosed in square brackets but ipv4 not.

# Postscript

I'm sorry for that I'm a beginner of GoLang. Please let me know if there is any bad practice, thanks for your patiences.

I'm still learning to write unit tests with GoLang. So there is no unit test for now, but I've already checked that it works fine in my production workflow.

### CODE REVIEW

File: proxy/http/client.go [hunk 1]
```diff
@@ -139,8 +148,42 @@ func (c *Client) Process(ctx context.Context, link *transport.Link, dialer inter
...
+	if len(header) == 0 {
+		return header, nil
+	}
+
+	inbound := session.InboundFromContext(ctx)
+	outbound := session.OutboundFromContext(ctx)
+
+	data := struct {
+		Source net.Destination
+		Target net.Destination
```
Comment (yuhan6665) [inline proxy/http/client.go:162]: It seems not used in your examples. Is target needed? I feel the receiving end would have it anyway

**Review by yuhan6665 — COMMENTED** [review]
Thanks for you work! Left a few comments.
I'm beginner in Golang as well. We can learn from each other :)

File: proxy/http/config.proto [hunk 2]
```diff
@@ -21,8 +21,14 @@ message ServerConfig {
...
+message Header {
+  string key = 1;
+  string value = 2;
+}
+
 // ClientConfig is the protobuf config for HTTP proxy client.
 message ClientConfig {
   // Sever is a list of HTTP server addresses.
   repeated xray.common.protocol.ServerEndpoint server = 1;
+  repeated Header header = 2;
```
[the remark this answers is not in the document]
Comment (PMExtra) [reply proxy/http/config.proto:33]: `xray.transport.internet.headers.http.Header` is designed to pick a random value for a key.
I referenced from `xray.transport.internet.websocket.Config` to make the current solution.

File: proxy/http/client.go [hunk 3]
```diff
@@ -30,6 +32,7 @@ import (
 type Client struct {
 	serverPicker  protocol.ServerPicker
 	policyManager policy.Manager
+	header        []*Header
```
[the remark this answers is not in the document]
Comment (PMExtra) [reply proxy/http/client.go:35]: According to `xray.transport.internet.websocket.Config` and `xray.transport.internet.http.Config`, the `Header` property is always in singular form. But in `xray.infra.conf.*` they are plural form. So I think it should be plural only in config file but singular in other codes.

File: proxy/http/client.go (hunk 1 above)
Comment (PMExtra) [reply proxy/http/client.go:162]: Yes, it's not needed in my case. I just want to make it more extendable for other usages.

**Comment by yuhan6665** [comment]
感谢 pr！

**Pull request merged**